tsfc/driver.py

```diff
@@ -396,7 +396,12 @@ def __call__(self, point_set, derivative=0):
                 from ufl.differentiation import ReferenceGrad
                 # from ufl.operators import grad
                 config = kernel_cfg.copy()
-                config.update(point_set=point_set)
+                if isinstance(point_set.expression, gem.Variable):
+                    # config for fem.GemPointContext
+                    config.update(point_indices=point_set.indices, point_expr=point_set.expression)
+                else:
+                    # config for fem.PointSetContext
+                    config.update(point_set=point_set)
```

Comment:
The fact that you have to repeat this pattern four times should warn you (code smell!) that you probably should not masquerade your point expressions as point sets in the first place...

Reply:
Not my code - this needs to be rebased onto `dual_base_fiat` again - sorry for the confusion. See https://github.com/firedrakeproject/tsfc/pull/218/

Reply:
Oh wait I see what you are talking about now - I hadn't quite realised that this code (again - not actually mine)  is creating `PointSet`s every time. I think just creating a `ContextBase` to get the core `config`, then updating `kwargs` as appropriate might be the way to go? Or I can do more extensive checking of the `**kwargs` argument in `fem.compile_ufl`

Reply:
Nope. I'm pretty sure @miklos1 is worried about the if isinstance repetition. That you have to update the config is probably fine. That you have to special case is a sign that the interface is badly designed.

Reply:
Yes, Lawrence is right.  Now, I cannot tell what the proper refactoring is without properly thinking the problem through, but this symptom is a clear sign, that something is wrong here.

Reply:
See up to date diff now that I've rebased onto the refactored  #218 (which is itself now based on the recent refactors you made @miklos1 ) - there's only one `isinstance` right now so the interface might not be the worst.